File: java/pe004.java
//file: pe004.java

public class pe004 {
  private static boolean is_palindromic(String number) {
    for (int i = 0; i < number.length(); i++) {
      if (number.charAt(i) != number.charAt(number.length() - (i + 1))) {
        return false;
      }
    }
    return true;
  }

  public static void main( String[] args ) {
    final int MAX3 = 999;
    final int MIN3 = 100;

    int maxp = 0;
    int mini = MIN3;
    int minj = MIN3;
    int maxj = MIN3;
    for (int i = MAX3; i >= mini; i--) {
      for (int j = maxj; j >= minj; j--) {
        int product = i * j;
        if (is_palindromic(Integer.toString(product))) {
          if (product > maxp) {
            maxp = product;
            mini = minj = product / 1000; /* optimize */
          }
        }
      }
      maxj = i;
    }
    System.out.println(maxp);
  }
}

File: typescript/pe004.ts
function is_palindromic(digits:Array<string>): boolean {
  let len:number = digits.length

  for (let i=0;i<len;i++) /* only really need to test half of the string - if even and int(half)+1 if odd  - but more complexity */
    if (digits[i] != digits[len-(i+1)])
      return false

  return true
}

function pe004(imin:number, imax:number, maxdigits:number): number {
  let maxp:number = 0
  let mini:number = imin
  let minj:number = imin
  let maxj:number = imax

  for (let i = imax; i >= mini; i--)
  {
    for (let j = maxj; j >= minj; j--)
    {
      let product:number = i * j
      let digits:Array<string> = product.toString().split("", maxdigits)
      if (is_palindromic(digits))
        if (product > maxp)
        {
          maxp = product
          mini = minj = product / 1000 // optimize
        }
    }
    maxj = i
  }
  return maxp
}

const MAX3:number = 999;
const MIN3:number = 100;
const MAXDIGITS:number = 6;

console.log(pe004(MIN3, MAX3, MAXDIGITS));
